### PR TITLE
Add class for embedding data on manifolds

### PR DESCRIPTION
**Pull request by Hzaatiti**
Add a class for embedding data on manifolds and unit tests.

should we delete the old example learning_graph_structured_data_h2.py ?

### CODE REVIEW

File: geomstats/learning/embedding_data.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        Maximum number of iterations for embedding.
+    lr : int
+        Learning rate for embedding.
+    n_negative : int
+        Number of negative samples to consider.
+    context_size : int
+        Size of the context size to consider.
+    """
+
+    data = None
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: True these are old C++ habits...

File: tests/test_embedding.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_grad_log_sigmoid(self):
+        """Test grad_log_sigmoid."""
+        point = gs.array([0.1, 0.3])
+        result = self.embedding_class.grad_log_sigmoid(point)
+
+        expected = gs.array([0.47502081, 0.42555748])
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply tests/test_embedding.py:46]: gs.assignment with tf backend for indices having a single value needs repairing still

File: tests/test_embedding.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        loss_value, loss_grad = self.embedding_class.loss(
+            point, point_context, point_negative)
+
+        expected_loss = 1.00322045
+        expected_grad = gs.array([[-0.16565083, -0.16565083]])
+
+        self.assertAllClose(loss_value, expected_loss, rtol=1e-3)
+        self.assertAllClose(loss_grad, expected_grad, rtol=1e-3)
+
+    @geomstats.tests.np_only
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply tests/test_embedding.py]: Embedding works now for pytorch :)

File: geomstats/learning/embedding_data.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
+"""Learning embeddings of data on manifolds."""
+
+import logging
+
+import geomstats.backend as gs
+from geomstats.geometry.poincare_ball import PoincareBall
+
+
+class Embedding ():
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: I went with 3. for the moment and have parameters set by default

[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: would u prefer option 2 for better showcasing the example ? although the code to embed is quite short now 
```
    embedding_class = HyperbolicEmbedding(karate_graph)

    embeddings = embedding_class.embed()
```

File: tests/test_embedding.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+"""Unit tests for embedding data class."""
+
+import geomstats.backend as gs
+import geomstats.tests
+from geomstats.datasets.utils import load_karate_graph
+from geomstats.geometry.poincare_ball import PoincareBall
+from geomstats.learning.embedding_data import Embedding
+
+
+class TestEmbedding(geomstats.tests.TestCase):
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply tests/test_embedding.py]: Should it be TestHyperbolicEmbedding now ?

File: geomstats/learning/embedding_data.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    dim : int
+        Number of dimensions of the manifold.
+    max_epochs : int
+        Maximum number of iterations for embedding.
+    lr : int
+        Learning rate for embedding.
+    n_negative : int
+        Number of negative samples to consider.
+    context_size : int
+        Size of the context size to consider.
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: Not sure if what I wrote now clarifies better ?

File: geomstats/learning/embedding_data.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+import logging
+
+import geomstats.backend as gs
+from geomstats.geometry.poincare_ball import PoincareBall
+
+
+class Embedding ():
+    """A class for learning embeddings of data on manifolds.
+
+    Currently supports embedding of graph-structured data on
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: No, as it is today it would need rework to generalize it to other manifolds, changed data to graph and set manifold as PoincareBall by default, only dim is given to the constructor.

File: geomstats/learning/embedding_data.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    manifold : object
+        An instance of a manifold object.
+    dim : int
+        Number of dimensions of the manifold.
+    max_epochs : int
+        Maximum number of iterations for embedding.
+    lr : int
+        Learning rate for embedding.
+    n_negative : int
+        Number of negative samples to consider.
```
[the remark this answers is not in the document]
Comment (Hzaatiti) [reply geomstats/learning/embedding_data.py]: ditto, not sure if the explanation now clarifies better?

File: tests/test_embedding.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ninamiolane) [reply tests/test_embedding.py:46]: @pchauchat any idea on this?

File: examples/learning_graph_embedding_and_predicting.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    Learns Poincaré Ball embedding by using Riemannian
+    gradient descent algorithm. Then K-means is applied
+    to learn labels of each data sample.
+    """
+    gs.random.seed(1234)
+
+    karate_graph = load_karate_graph()
+
+    hyperbolic_embedding = HyperbolicEmbedding(karate_graph)
```
Comment (ninamiolane) [inline examples/learning_graph_embedding_and_predicting.py:25]: It would be better with a syntax that follows scikit-learn's logic.
```
hyperbolic_embedding = HyperbolicEmbedding()
embeddings = hyperbolic_embedding.embed(karate_graph)
```
the object hyperbolic_embedding should not depend on the dataset that will be passed to it, right? It could theoretically be used to embed another dataset later?

File: tests/test_embedding.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ninamiolane) [reply tests/test_embedding.py]: Both the test filename and the test's class should match the module being tested. Here, the module being tested is `prepare_graph_data.py`. Thus:
- `test_prepare_graph_data.py`
- `TestPrepareGraphData`
